lib/common/tests/rules/pcmk__evaluate_date_expression_test.c

```diff
@@ -0,0 +1,1061 @@
+}
+
+#define EXPR_GT_INVALID_START                           \
+    "<" PCMK_XE_DATE_EXPRESSION " " PCMK_XA_ID "='e' "  \
+    PCMK_XA_OPERATION "='" PCMK_VALUE_GT "' "           \
+    PCMK_XA_START "='not-a-datetime' />"
+
+static void
+gt_invalid_start(void **state)
+{
+    xmlNodePtr xml = string2xml(EXPR_GT_INVALID_START);
+    crm_time_t *t = NULL;
+
+    t = crm_time_new("2024-01-01");
+    assert_int_equal(pcmk__evaluate_date_expression(xml, t, NULL),
+                     pcmk_rc_undetermined);
+    crm_time_free(t);
+
+    free_xml(xml);
+}
+
+#define EXPR_GT_VALID                                   \
+    "<" PCMK_XE_DATE_EXPRESSION " " PCMK_XA_ID "='e' "  \
+    PCMK_XA_OPERATION "='" PCMK_VALUE_GT "' "           \
+    PCMK_XA_START "='2024-02-01 12:00:00' />"
+
+static void
+gt_valid(void **state)
+{
+    xmlNodePtr xml = string2xml(EXPR_GT_VALID);
+    crm_time_t *t = NULL;
+    crm_time_t *next_change = NULL;
+    crm_time_t *reference = NULL;
+
+    // t is before start, next change is before start
+    t = crm_time_new("2024-01-01 04:30:05");
+    next_change = crm_time_new("2024-01-01 11:00:00");
+    reference = crm_time_new("2024-01-01 11:00:00");
+    assert_int_equal(pcmk__evaluate_date_expression(xml, t, next_change),
+                     pcmk_rc_before_range);
+    assert_int_equal(crm_time_compare(next_change, reference), 0);
+    crm_time_free(reference);
+    crm_time_free(next_change);
+    crm_time_free(t);
+
+    // t is before start, next change is after start
+    t = crm_time_new("2024-02-01 11:59:59");
+    next_change = crm_time_new("2024-02-01 18:00:00");
+    reference = crm_time_new("2024-02-01 12:00:01");
+    assert_int_equal(pcmk__evaluate_date_expression(xml, t, next_change),
+                     pcmk_rc_before_range);
+    assert_int_equal(crm_time_compare(next_change, reference), 0);
+    crm_time_free(reference);
+    crm_time_free(next_change);
+    crm_time_free(t);
+
+    // t is equal to start, next change is after start
+    t = crm_time_new("2024-02-01 12:00:00");
+    next_change = crm_time_new("2024-02-01 18:00:00");
+    reference = crm_time_new("2024-02-01 12:00:01");
+    assert_int_equal(pcmk__evaluate_date_expression(xml, t, next_change),
+                     pcmk_rc_before_range);
+    assert_int_equal(crm_time_compare(next_change, reference), 0);
+    crm_time_free(reference);
+    crm_time_free(next_change);
+    crm_time_free(t);
+
+    // t is one second after start, next change is after start
+    t = crm_time_new("2024-02-01 12:00:01");
+    next_change = crm_time_new("2024-02-01 18:00:00");
+    reference = crm_time_new("2024-02-01 18:00:00");
+    assert_int_equal(pcmk__evaluate_date_expression(xml, t, next_change),
+                     pcmk_rc_within_range);
+    assert_int_equal(crm_time_compare(next_change, reference), 0);
+    crm_time_free(reference);
+    crm_time_free(next_change);
+    crm_time_free(t);
+
+    // t is after start, next change is after start
```

Comment:
I don't think this gives us any information that the "one second after start, next change is after start" test doesn't give us, but it doesn't hurt

Reply:
Yeah, just paranoid